src/libCli/Completion.cpp

```diff
@@ -19,198 +19,242 @@ using namespace ArgParse;
+    {
+        // completion requested :)
+        if (f_debug)
         {
-          printf("nospace! cand='%s', n=%zu, start=%zu, end = %zu\n",candidateStr.c_str(), n, start, end);
+            std::cerr << "Input string \"" << f_args << std::endl;
         }
 
-        suggestion = candidateStr.substr(start, std::string::npos);
-        size_t trimEnd = suggestion.find_last_not_of(' ');
-        suggestion = suggestion.substr(0, trimEnd+1);
-        //suggestion = candidateStr.substr(start, end-start+1);
+        // List with all suggestions for this candidate. Used to eliminate duplicate suggestions
+        std::vector<std::string> suggestions;
 
-        if(!suggestionDoc.empty())
+        // size_t n = parseTree.getMatchedString().size();
+        size_t n = f_args.size();
+        for (auto candidate : f_candidates)
         {
-            suggestion = suggestion + "\t" + suggestionDoc;
-        }
+            std::string candidateStr = candidate->getMatchedStringRaw();
 
-        if(f_debug)
-        {
-            printf("post: '%s'\n", suggestion.c_str());
+            if (f_debug)
+            {
+                std::cout << "------------------------------------------------" << std::endl;
+                std::cout << "candiate string: " << candidateStr << std::endl;
+                std::cout << "candiate's debug string: " << std::endl;
+                std::cout << candidate->getDebugString() << std::endl;
+                std::cout << "*************Suggestion****************";
+                std::cout << "------------------------------------------------" << std::endl;
+            }
+
+            std::string suggestion;
+            size_t start = n;
+            size_t end;
+
+            std::string suggestionDoc = candidate.get()->getShortDocument();
+            std::string suggestionDocRoot = f_parseTree.getShortDocument();
+            // here the document (description of the tab completion) extracted from the original parsed tree and from the candidate tree should be compared,
+            // if there is no new document of candidate in the parsed tree, that means the both extracted documents must same,
+            // we only find the document at the right most position in the parsed tree (the newest) for tab completion.
+            if (suggestionDocRoot == suggestionDoc)
+            {
+                suggestionDoc = "";
+            }
+
+            if (f_debug)
+            {
+                std::cout << "suggestionDoc: " << suggestionDoc << std::endl;
+                printf("pre: '%s'\n", candidateStr.c_str());
+                printf("candidateStr[n=%zu] = '%c'\n", n, candidateStr[n]);
+            }
+
+            bool isTrimmed = false;
+            suggestion = getLastArgument(candidateStr, " ", f_args, isTrimmed);
+
+            if (suggestions.empty())
+            {
+                suggestions.push_back(suggestion);
+            }
+            else if (suggestion == suggestions.back())
```

Comment:
this assumes that duplicates always come directly after each other. is this a true assumption?
If so you do not need the vector at all, can just store last completion and compare.